### PR TITLE
Slightly refactor daemonize.c

### PR DESCRIPTION
**Pull request by gperciva**

### CODE REVIEW

**Comment by gperciva** [comment]
This includes #467.

I'm fairly confident that `noeintr_readbyte()` is a good clarifying addition.  Less confident about `writepid()`... I found it useful to abstract that away when I was trying to understand this code, but you might prefer to keep it in `daemonize()` itself.

File: util/daemonize.c [hunk 1]
```diff
@@ -7,6 +7,41 @@
 
 #include "daemonize.h"
 
+/* Read one byte from ${fd}, looping upon EINTR.  Do not store that byte. */
+static int
+noeintr_readbyte(int fd)
```
Comment (cperciva) [inline util/daemonize.c:12]: Since we have a `noeintr.c`, we shouldn't have functions named `noeintr_*` in other files.

File: util/daemonize.c [hunk 2]
```diff
@@ -53,29 +88,19 @@ daemonize(const char * spid)
 			warnp("close");
 			goto err1;
 		}
-		do {
-			switch (read(fd[0], &dummy, 1)) {
```
Comment (cperciva) [inline util/daemonize.c:57]: I think we can get rid of the dummy variable now?

Comment (gperciva) [reply util/daemonize.c:57]: Unfortunately it's needed for writing to `fd`:

```
        /* Tell the parent to suicide. */
        if (noeintr_write(fd[1], &dummy, 1) == -1) {
                warnp("write");
                goto die;       
        }
```

I was very tempted to move that `noeintr_write()` into a separate function (maybe called `sendbyte()`?), to get rid of `dummy` and to be semantically clearer, but I figured that would be over-engineering it.

File: util/daemonize.c (hunk 1 above)
Comment (gperciva) [reply util/daemonize.c]: done

File: util/daemonize.c [hunk 3]
```diff
@@ -62,6 +62,10 @@ daemonize(const char * spid)
 				/* EOF -- the child died without poking us. */
 				goto err1;
 			case 1:
+				/* We don't need to read from this any more. */
```
Comment (gperciva) [inline util/daemonize.c]: I added this in #467, which you haven't commented on.

(I just wanted to make sure that you're fully aware of this change)

File: util/daemonize.c [hunk 4]
```diff
@@ -7,6 +7,89 @@
...
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Send a byte to ${fd}, ignoring EINTR. */
+static int
+sendbyte(int fd)
```
Comment (gperciva) [inline util/daemonize.c:47]: I added this as a `MAYBE???` commit.

I'd actually prefer to include this change, but I'm betting that you'll say that the (source-code) overhead of `sendbyte()` isn't worth it.

File: util/daemonize.c (hunk 2 above)
Comment (cperciva) [reply util/daemonize.c:57]: Right, I forgot about that.  Yeah no point moving that code.

File: util/daemonize.c (hunk 4 above)
Comment (cperciva) [reply util/daemonize.c:47]: Yeah, not worth it.

**Comment by gperciva** [comment]
Rebased.

File: util/daemonize.c [hunk 5]
```diff
@@ -7,6 +7,69 @@
 
 #include "daemonize.h"
 
+/* Read one byte from ${fd}, looping upon EINTR.  Do not store that byte. */
```
Comment (cperciva) [inline util/daemonize.c:10]: "Read and discard one byte from ${fd}, looping upon EINTR."

**Comment by cperciva** [comment]
Good aside from minor rewording of a comment.

**Comment by gperciva** [comment]
Rebased.  I also added "and discard" to the comment where we're calling `readbyte()`:

```
/* Read and discard a byte from the read end of the pipe. */
if (readbyte(fd[0]))
```

**Pull request merged**